Base/QTCore/qSlicerExtensionsManagerModel.cxx

```diff
@@ -812,6 +822,145 @@ void qSlicerExtensionsManagerModelPrivate::saveExtensionDescription(
+  QStringList historyList = settings.value("ExtensionsHistory/Revisions/" + this->SlicerRevision).toStringList();
+  for (int i = 0; i < scheduledForRemovalList.length(); i++)
+  {
+    historyList.removeOne(scheduledForRemovalList.at(i));
+  }
+  historyList.removeDuplicates();
+  settings.setValue("ExtensionsHistory/Revisions/" + this->SlicerRevision, historyList);
+  settings.setValue("ExtensionsHistory/ScheduledForRemoval", "");
+}
+
+// --------------------------------------------------------------------------
+QVariantMap qSlicerExtensionsManagerModelPrivate::getExtensionsInfoFromPreviousInstallations(
+  const QString& extensionsHistorySettingsFile)
+{
+  Q_Q(qSlicerExtensionsManagerModel);
+  QVariantMap extensionsHistoryInformation;
+  QSettings settings(extensionsHistorySettingsFile, QSettings::IniFormat);
+  settings.beginGroup("ExtensionsHistory/Revisions");
+  QStringList revisions = settings.childKeys();
+  revisions.sort();	//revisions sorted ascending
+  int lastRevision = -1;
+  for (int i = revisions.length() - 1; i >= 0; i--)		//the revision with the highest number not equal the current one is considered to be the last revision
+  {
+    if (revisions[i] != this->SlicerRevision)
+    {
+      lastRevision = i;
+      break;
+    }
+  }
+  if (lastRevision == -1)
+  {
+    return extensionsHistoryInformation;
+  }
+
+  for (int i = 0; i < revisions.length(); i++)
+  {
+    QVariantMap curExtensionInfo;
+    const QStringList& extensionNames = settings.value(revisions.at(i)).toStringList();
+    for (int j = 0; j < extensionNames.length(); j++)
+    {
+      QString extensionName = extensionNames.at(j);
+      QString extensionId = "";
+
+      //make sure that no invalid white space leads to invalid extension names
+      //also catches empty entries in [ExtensionHistory] in Slicer.ini
+      extensionName = extensionName.simplified();
+      extensionName.replace(" ", "");
+      if (extensionName.length() == 0)
+      {
+        break;
+      }
+
+      curExtensionInfo.insert("UsedLastInRevision", revisions.at(i));
+      if (i == lastRevision || (extensionsHistoryInformation.contains(extensionName) &&
+        extensionsHistoryInformation.value(extensionName).toMap().value("WasInstalledInLastRevision").toBool()))
+      {
+        curExtensionInfo.insert("WasInstalledInLastRevision", true);
+      }
+      else
+      {
+        curExtensionInfo.insert("WasInstalledInLastRevision", false);
+      }
+      curExtensionInfo.insert("IsInstalled", q->isExtensionInstalled(extensionName));
+      bool isCompatible = true;
+
+      if (!q->isExtensionInstalled(extensionName)) {
+        qMidasAPI::ParametersType parameters;
+        parameters["productname"] = extensionName;
+        parameters["slicer_revision"] = q->slicerRevision();
+        parameters["os"] = q->slicerOs();
+        parameters["arch"] = q->slicerArch();
+        const ExtensionMetadataType& metaData = retrieveExtensionMetadata(parameters);
```

Comment:
There are still two issues when retrieving extension metadata remotely:
1. the `Q_ASSERT` in `qMidasAPI::synchronousQuery` is triggered if an invalid extension name is provided to `qSlicerExtensionsManagerModelPrivate::retrieveExtensionMetadata` (as the source for the extension name is the `[ExtensionsHistory]` section in `Slicer.ini`, this is a possible scenario, if e.g. the file was somehow edited)
It seems that an invalid extension name does not set `queryResult.Error.isNull() == true` and therefor the assert is triggered, as this results in `queryResult.Result.count() == 0`
2. If no internet connectivity is available, the timeout seems not to trigger. `qMidasAPI::synchronousQuery` has a standard timeout of 2500 msec, however, it seems that no timeout error is fired ( `QObject::connect(&restAPI, SIGNAL(errorReceived(QUuid,QString)), &eventLoop, SLOT(quit()));`) as Slicer gets stuck in the event processing loop (`eventLoop.exec()`)